substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/SubstrateJVM.java

```diff
@@ -311,11 +315,29 @@ public static long getThreadId(Thread thread) {
     @Uninterruptible(reason = "Called from uninterruptible code.", mayBeInlined = true)
     public static long getCurrentThreadId() {
         if (HasJfrSupport.get()) {
+            maybeRegisterVirtualThread(Thread.currentThread());
             return JavaThreads.getCurrentThreadId();
         }
         return 0;
     }
 
+    /**
+     * Register virtual threads if they aren't registered already. Platform threads are registered
+     * eagerly when started and at chunk rotations.
+     */
+    @Uninterruptible(reason = "Epoch should not change while checking generation.")
+    private static void maybeRegisterVirtualThread(Thread thread) {
```

Comment:
Doesn't this break once we leave uninterruptible code?
- we start in epoch 0
- thread A calls `getThreadId(...)`, which registers a `vthread` for epoch 0 in `maybeRegisterVirtualThread`
- thread A leaves the uninterruptible code and gets blocked at a safepoint
- thread B changes the epoch from 0 to 1
- after the safepoint, thread A continues execution and uses the thread id that was returned by `getThreadId(...)` even though the vthread was only registered for epoch 0

One example where I think that this might happen: `JavaMonitorQueuedSynchronizer.ConditionNode.notifierJfrTid`

Reply:
I think the situation you describe is actually ok.  Once the epoch changes, new event emissions from "thread A" will fail the epoch generation check, and registration will be re-done. Since `getThreadID()` is called for every event emission, the important thing is that gathering event info with `getThreadID()` and the writing of the event data to the JFR buffer are done within the same block of uninterruptible code, which does indeed happen (ex. all `emit0` methods). 

Calls to `getThreadID()` outside of the event emission path (ex. with `JavaMonitorQueuedSynchronizer.ConditionNode.notifierJfrTid`) result in unnecessary registration checks, but shouldn't break anything.

Reply:
Ok, so you essentially assume that `getThreadID()` will be called again for thread A once the thread ID is actually used in a JFR event. However, I don't think that this assumption is valid. Here is one example:
- Thread B blocks at `ConditionNode`.
- Thread A notifies thread B and sets `ConditionNode.notifierJfrTid`.
- The JFR epoch changes.
- Once thread B wakes up, it uses the value of `ConditionNode.notifierJfrTid` (i.e., the ID of thread A) when emitting a `JavaMonitorWaitEvent`. However, thread A is not registered for the current epoch.

Reply:
Ok I see. I mistakenly thought we always used `JfrNativeEventWriter.putThread` when dealing with threads. But in that case we use a the plain TID directly. 

I think we'll have to switch to always using `putThread` whenever dealing with the JFR `Thread` type.  What do you think of this?

Reply:
The downside is that we'll have to remember to give careful consideration to emitting  JFR `Thread` data. 
I'm not sure there's a better place to put the registration check though. And we'll need to check at some point in order to copy Hotspot and do lazy registration.

Reply:
I've just pushed a commit to update all the places we write thread IDs as JFR event data to use either `putThread(Thread)` or `putThread(long)`. This should ensure that threads are always registered if they get referenced from event data. The downside is that we need to remember to use `putThread`. 

Another unfortunate thing is that we can only do checks with `putThread(Thread)` because with only the TID we have no way of knowing whether the thread is virtual and cannot access `Target_java_lang_VirtualThread.jfrGeneration`.